crates/task-impls/src/helpers.rs

```diff
@@ -502,9 +503,29 @@ pub async fn validate_proposal_safety_and_liveness<
     // Create a positive vote if either liveness or safety check
     // passes.
 
-    // Liveness check.
     {
         let consensus_reader = validation_info.consensus.read().await;
+        // Epoch safety check:
+        // The proposal is safe if
+        // 1. the proposed block and the justify QC block belong to the same epoch or
+        // 2. the justify QC is the eQC for the previous block
+        let proposal_epoch =
```

Comment:
Do we also need all this logic in the `validate_proposal_liveness` path, I suppose we can't actually do it there because we are missing the previous proposal but don't we need something to check the epochs?

Reply:
I don't think so, I was just trying to be sure we didn't. I also don't think we can do the check in the liveness path, and we won't vote in that case either